File: fenwick_tree.rs
struct FenwickTree {
    bit: Vec<i32>,
}

impl FenwickTree {
    fn with_capacity(n: usize) -> Self {
        Self { bit: vec![0; n] }
    }

    fn sum(&self, mut r: usize) -> i32 {
        let mut ret = 0;
        loop {
            ret += self.bit[r];
            r &= r + 1;
            if r == 0 { break; }
            r -= 1;
        }
        ret
    }

    fn sum_range(&self, l: usize, r: usize) -> i32 {
        let mut ret = self.sum(r);
        if l > 0 { ret -= self.sum(l - 1); }
        ret
    }

    fn add(&mut self, mut idx: usize, delta: i32) {
        while idx < self.bit.len() {
            self.bit[idx] += delta;
            idx |= idx + 1;
        }
    }
}

File: fenwick_tree.cpp
struct FenwickTree {
    vector<int> bit;

    FenwickTree(int n) {
        bit.assign(n, 0);
    }

    int sum(int r) {
        int ret = 0;
        for (; r >= 0; r = (r & (r + 1)) - 1) {
            ret += bit[r];
        }
        return ret;
    }

    int sum(int l, int r) {
        int ret = sum(r);
        if (l >= 1) ret -= sum(l - 1);
        return ret;
    }

    void add(int idx, int delta) {
        for (; idx < bit.size(); idx |= (idx + 1)) {
            bit[idx] += delta;
        }
    }
};
